src/wrapper/clap-juce-wrapper.cpp

```diff
@@ -819,176 +824,219 @@ class ClapJuceWrapper : public clap::helpers::Plugin<
-                     * to send it back out as a UI message but we do want to trigger any *other*
-                     * listeners which may be attached. So suppress my listeners while we send this
-                     * event.
-                     */
-                    auto g = AtomicTGuard<bool>(supressParameterChangeMessages, true);
-                    paramSetValueAndNotifyIfChanged(*jp, (float)nf);
-                }
-                break;
-                case CLAP_EVENT_PARAM_MOD:
-                {
-                }
-                break;
-                case CLAP_EVENT_NOTE_END:
-                {
-                    // Why do you send me this, Alex?
-                }
-                break;
-                default:
-                {
-                    DBG("Unknown message type " << (int)evt->type);
-                    // In theory I should never get this.
-                    // jassertfalse
-                }
-                break;
-                }
-            }
+            auto event = events->get(events, 0);
+            nextEventTime = (int)event->time;
         }
 
-        // We process in place so
-        static constexpr uint32_t maxBuses = 128;
-        std::array<float *, maxBuses> busses{};
-        busses.fill(nullptr);
+        auto processEvent = [&](int sampleOffset) {
+            auto event = events->get(events, (uint32_t)currentEvent);
+            process_clap_event(event, sampleOffset);
 
-        /*DBG("IO Configuration: I=" << (int)process->audio_inputs_count << " O="
-                                   << (int)process->audio_outputs_count << " MX=" << (int)mx);
-        DBG("Plugin Configuration: IC=" << processor->getTotalNumInputChannels()
-                                        << " OC=" << processor->getTotalNumOutputChannels());
-        */
+            currentEvent++;
+            nextEventTime = (currentEvent < numEvents)
+                                ? (int)events->get(events, (uint32_t)currentEvent)->time
+                                : numSamples;
+        };
 
         /*
          * OK so here is what JUCE expects in its audio buffer. It *always* uses input as output
          * buffer so we need to create a buffer where each channel is the channel of the associated
          * output pointer (fine) and then the inputs need to either check they are the same or copy.
          */
+        static constexpr uint32_t maxBuses = 128;
+        std::array<float *, maxBuses> busses{};
+        busses.fill(nullptr);
 
-        /*
-         * So first lets load up with our outputs
-         */
-        uint32_t ochans = 0;
-        for (uint32_t idx = 0; idx < process->audio_outputs_count && ochans < maxBuses; ++idx)
+        for (int n = 0; n < numSamples;)
         {
-            for (uint32_t ch = 0; ch < process->audio_outputs[idx].channel_count; ++ch)
+            while (nextEventTime == n) // process all events at this timestep
```

Comment:
Why do we need this here and also the while below?

Reply:
So in order to know how many samples to process, we need to know when is the next timestamp that has an event. In order to know that next timestamp, we need to process all of the events at the current timestamp. I'll add some comments to make the reasoning more clear.